moveos/moveos-stdlib/rooch-framework/sources/account.move

```diff
@@ -46,24 +99,213 @@ module rooch_framework::account{
+      );
+      let signer = create_account_unchecked(addr);
+      let signer_cap = SignerCapability { addr };
+      (signer, signer_cap)
+   }
+
+
+   /// Return the current sequence number at `addr`
+   public fun sequence_number(addr: address): u64 acquires Account {
+      sequence_number_for_account(borrow_global<Account>(addr))
+   }
+
+   public(friend) fun increment_sequence_number(addr: address) acquires Account {
+      let sequence_number = &mut borrow_global_mut<Account>(addr).sequence_number;
+
+      assert!(
+         (*sequence_number as u128) < MAX_U64,
+         error::out_of_range(ESequenceNumberTooBig)
+      );
+
+      *sequence_number = *sequence_number + 1;
+   }
+
+   /// Helper to return the sequence number field for given `account`
+   fun sequence_number_for_account(account: &Account): u64 {
+      account.sequence_number
+   }
+
+   /// Return the current TokenType balance of the account at `addr`.
+   public fun balance<TokenType: store>(_addr: address): u128 {
+      //TODO token standard, with balance precesion(u64|u128|u256)
+      0u128
+   }
+
+   #[view]
+   public fun get_authentication_key(addr: address): vector<u8> acquires Account {
+      *&borrow_global<Account>(addr).authentication_key
+   }
+
+   public fun signer_address(cap: &SignerCapability): address {
+      cap.addr
+   }
+
+   public fun is_resource_account(addr: address): bool {
+      exists<ResourceAccount>(addr)
+   }
+
+
    #[view]
    public fun exists_at(addr: address): bool {
       exists<Account>(addr)
    }
 
 
    native fun create_signer(addr: address): signer;
+
+
+   /// A resource account is used to manage resources independent of an account managed by a user.
+   /// In Rooch a resource account is created based upon the sha3 256 of the source's address and additional seed data.
+   /// A resource account can only be created once
+   public fun create_resource_account(source: &signer): (signer, SignerCapability) acquires Account {
+      let source_addr = signer::address_of(source);
+      let seed = generate_seed_bytes(&source_addr);
+      let resource_addr = create_resource_address(&source_addr, seed);
+      let resource_signer = if (exists_at(resource_addr)) {
+         let account = borrow_global<Account>(resource_addr);
+         assert!(account.sequence_number == 0, error::invalid_state(EResourceAccountAlreadyUsed));
+         create_signer(resource_addr)
+      } else {
+         create_account_unchecked(resource_addr)
+      };
+
+      // By default, only the SignerCapability should have control over the resource account and not the auth key.
+      // If the source account wants direct control via auth key, they would need to explicitly rotate the auth key
+      // of the resource account using the SignerCapability.
+      rotate_authentication_key_internal(&resource_signer, ZERO_AUTH_KEY);
+
+      assert!(!is_resource_account(resource_addr), error::invalid_state(EAccountIsAlreadyResourceAccount));
```

Comment:
Is it better to move the `assert` statement before creating the signer? 
fail fast